File: back-end/mercado-inteligente-spring/src/main/java/br/edu/usf/mercadointeligentespring/gateways/database/UserGatewayDatabase.java
package br.edu.usf.mercadointeligentespring.gateways.database;

import br.edu.usf.mercadointeligentespring.gateways.UserGateway;
import br.edu.usf.mercadointeligentespring.gateways.database.repository.UserRepository;
import br.edu.usf.mercadointeligentespring.gateways.dto.UserDTO;
import java.util.List;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
class UserGatewayDatabase implements UserGateway {

    private final UserRepository userRepository;

    @Override
    public UserDTO create(UserDTO user) {
        return userRepository.save(user);
    }

    @Override
    public UserDTO update(Long id, UserDTO userDTO) {
        userDTO.setId(id);
        return userRepository.save(userDTO);
    }

    @Override
    public void delete(Long id) {
        userRepository.deleteById(id);
    }

    @Override
    public Optional<UserDTO> findById(Long id) {
        return userRepository.findById(id);
    }

    @Override
    public List<UserDTO> findAll() {
        return (List<UserDTO>) userRepository.findAll();
    }
}

File: back-end/mercado-inteligente/src/main/kotlin/br/edu/usf/gateways/database/UserGatewayDatabase.kt
package br.edu.usf.gateways.database

import br.edu.usf.gateways.UserGateway
import br.edu.usf.gateways.database.repository.UserRepository
import br.edu.usf.gateways.dto.UserDTO
import javax.enterprise.context.ApplicationScoped
import javax.enterprise.inject.Default
import javax.inject.Inject

@ApplicationScoped
class UserGatewayDatabase : UserGateway {
    @Inject
    @field: Default
    lateinit var userRepository: UserRepository

    override fun create(user: UserDTO): UserDTO = userRepository.create(user)

    override fun update(id: Long, user: UserDTO) = userRepository.update(id, user)

    override fun delete(id: Long) = userRepository.deleteById(id)

    override fun findById(id: Long): UserDTO? {
        return userRepository.findById(id)
    }

    override fun findAll(): List<UserDTO> {
        return userRepository.findAll().list()
    }
}